### PR TITLE
[WIP] CleanupBackend

### PR DESCRIPTION
**Pull request by raimund-schluessler**
Cleanup Backend, implement own database mapper.

- [x] Refactor reminder parsing
- [x] Probably create own services for handling vtodo properties to not blow up ```TasksService``` and ```Helper```

### CODE REVIEW

File: service/tasksservice.php [hunk 1]
```diff
@@ -221,15 +199,9 @@ public function setPercentComplete($taskID, $percent_complete) {
...
-	public function setPriority($taskID, $priority){
-		$vcalendar = \OC_Calendar_App::getVCalendar($taskID);
-		$vtodo = $vcalendar->VTODO;
-		if($priority){
-			$vtodo->PRIORITY = (10 - $priority) % 10;
-		}else{
-			unset($vtodo->{'PRIORITY'});
-		}
-		return \OC_Calendar_Object::edit($taskID, $vcalendar->serialize());
+	public function priority($taskID, $priority){
```
Comment (Henni) [inline service/tasksservice.php]: @raimund-schluessler why did you change the name of this function?
All other functions are also called setXYZ().

File: service/helper.php [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+	/**
+	 * set property of a task
+	 * 
+	 * @param  int    $taskID
+	 * @param  string $property
+	 * @param  mixed  $value
+	 * @return bool
+	 * @throws \Exception
+	 */
+	public function setProperty($taskID,$property,$value){
```
Comment (Henni) [inline service/helper.php]: setProperty is only used in the tasks service. We could move it over there as a private function.

Comment (Henni) [inline service/helper.php]: I think every function in the helper class is just used by the tasks service.
We have to restructure the tasks service and helper functions I believe to remove this code smell.

File: service/tasksservice.php (hunk 1 above)
Comment (raimund-schluessler) [inline service/tasksservice.php]: Yes, indeed. I don't know, why I changed it. I think I will undo the change ;)

File: service/helper.php (hunk 2 above)
Comment (raimund-schluessler) [inline service/helper.php]: Yes, that is more or less what I meant in the first comment of the PR ;)

**Comment by raimund-schluessler** [comment]
- [ ] Unify ```Helper::formatDate``` and ```TaskParser::parseDateString```

**Comment by raimund-schluessler** [comment]
@Henni Mind having a look at this? I think it is good to merge now.

File: service/tasksservice.php [hunk 3]
```diff
@@ -392,17 +348,17 @@ public function removeCategory($taskID, $category){
...
+			// remove category
+			$key = array_search($category, $taskcategories);
+			if ($key !== null && $key !== false){
+				unset($taskcategories[$key]);
+				if(count($taskcategories)){
+					$vtodo->CATEGORIES = $taskcategories;
+				} else{
+					unset($vtodo->{'CATEGORIES'});
+				}
+				return \OC_Calendar_Object::edit($taskID, $vcalendar->serialize());
```
Comment (Henni) [inline service/tasksservice.php]: This function doesn't return anything in some cases. We should add a 'default return' after line 362.

File: service/helper.php [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+		}
+		return $t1 > $t2 ? 1 : -1;
+	}
+
+	/**
+	 * sort get comment by ID
+	 *
+	 * @param array $vtodo
+	 * @param string $commentID
+	 * @return array
```
Comment (Henni) [inline service/helper.php]: This function doesn't return an array. I think it returns an `int`.

File: service/reminderservice.php [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+			'action'	=> $reminderAction,
+			'date'		=> $reminderDate,
+			'duration'	=> $reminderDuration,
+		);
+	}
+
+	/**
+	 * parse reminder duration
+	 *
+	 * @param string $reminder
```
Comment (Henni) [inline service/reminderservice.php]: $reminder isn't of type string. It should be `mixed` or something more specific instead.

**Comment by raimund-schluessler** [comment]
@Henni I implemented all your suggested changes and rebased.

**Comment by Henni** [comment]
Awesome!
Then it's ready to merge in my opinion.

**Comment by raimund-schluessler** [comment]
Great, thanks for your feedback!
